diff --git a/samples/Thumbnail-Sai2.cpp b/samples/Thumbnail-Sai2.cpp
index 2cfba5e..5a3d929 100644
--- a/samples/Thumbnail-Sai2.cpp
+++ b/samples/Thumbnail-Sai2.cpp
@@ -31,15 +31,12 @@ using ThumbnailT
 	= std::tuple<std::unique_ptr<std::byte[]>, std::uint32_t, std::uint32_t>;
 
 bool IterateCanvasItem(
-	const std::filesystem::path& FilePath, const sai2::CanvasHeader& Header,
+	const std::filesystem::path& FilePath,
+	const std::filesystem::path& DestPath,
+	const sai2::CanvasHeader& Header,
 	const sai2::CanvasEntry& TableEntry, std::span<const std::byte> Bytes
 )
 {
-	std::printf(
-		"%.4s:%08X @ %016lX\n", reinterpret_cast<const char*>(&TableEntry.Type),
-		TableEntry.LayerID, TableEntry.BlobsOffset
-	);
-
 	switch( TableEntry.Type )
 	{
 	case sai2::CanvasDataType::ThumbnailLossy:
@@ -47,10 +44,6 @@ bool IterateCanvasItem(
 		if( const auto JpegStream = sai2::ExtractJssfToJpeg(Bytes);
 			!std::get<0>(JpegStream).empty() )
 		{
-			std::filesystem::path DestPath(FilePath);
-			DestPath.replace_filename(FilePath.stem().string() + "-thumbnail");
-			DestPath.replace_extension("png");
-
 			// Decode jpeg stream
 			const auto JpegData = std::get<0>(JpegStream);
 
@@ -133,43 +126,47 @@ int main(int argc, char* argv[])
 		Args.emplace_back(argv[i]);
 	}
 
-	for( const std::string_view& Arg : Args )
+	if(Args.size() < 2)
 	{
-		std::puts(Arg.data());
-		const std::filesystem::path FilePath(Arg);
-		if( !(std::filesystem::exists(FilePath)
-			  && std::filesystem::is_regular_file(FilePath)) )
-		{
-			// Not a file
-			std::printf("Invalid path %s\n", Arg.data());
-			continue;
-		}
-		const std::uintmax_t FileSize = std::filesystem::file_size(FilePath);
+		std::printf("Input and output paths required.\n");
+		return EXIT_FAILURE;
+	}
 
-		std::ifstream          File(FilePath);
-		std::vector<std::byte> FileData(FileSize);
+	const std::filesystem::path FilePath(Args[0]);
+	const std::filesystem::path DestPath(Args[1]);
+	if( !(std::filesystem::exists(FilePath)
+			&& std::filesystem::is_regular_file(FilePath)) )
+	{
+		// Not a file
+		std::printf("Invalid path %s\n", FilePath.string());
+		return EXIT_FAILURE;
+	}
+	const std::uintmax_t FileSize = std::filesystem::file_size(FilePath);
 
-		if( File.is_open() )
-		{
-			File.read(reinterpret_cast<char*>(FileData.data()), FileSize);
-			File.close();
-
-			const auto CanvasDataProc = [&FilePath](
-											const sai2::CanvasHeader& Header,
-											const sai2::CanvasEntry& TableEntry,
-											std::span<const std::byte> Bytes
-										) {
-				return IterateCanvasItem(FilePath, Header, TableEntry, Bytes);
-			};
-
-			sai2::IterateCanvasData(FileData, CanvasDataProc);
-		}
-		else
-		{
-			std::printf("Error reading file contents %s\n", Arg.data());
-			continue;
-		}
+	std::ifstream          File(FilePath);
+	std::vector<std::byte> FileData(FileSize);
+
+	if( File.is_open() )
+	{
+		File.read(reinterpret_cast<char*>(FileData.data()), FileSize);
+		File.close();
+
+		const auto CanvasDataProc = [&FilePath, &DestPath](
+										const sai2::CanvasHeader& Header,
+										const sai2::CanvasEntry& TableEntry,
+										std::span<const std::byte> Bytes
+									) {
+			return IterateCanvasItem(FilePath, DestPath, Header, TableEntry, Bytes);
+		};
+
+		sai2::IterateCanvasData(FileData, CanvasDataProc);
+	}
+	else
+	{
+		std::printf("Error reading file contents %s\n", FilePath.string());
+		return EXIT_FAILURE;
 	}
 
+
 	return EXIT_SUCCESS;
 }
\ No newline at end of file
